core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -3789,9 +3820,129 @@ public void validateInsert(SqlInsert insert) {
 
     checkTypeAssignment(logicalSourceRowType, logicalTargetRowType, insert);
 
+    checkConstraint(table, source, logicalTargetRowType);
+
     validateAccess(insert.getTargetTable(), table, SqlAccessEnum.INSERT);
   }
 
+  /**
+   * Validate insert values against the constraint of a modifiable view.
+   * @param table                The table wrapping the modifiable view.
+   * @param source               The values being inserted.
+   * @param logicalTargetRowType The target type.
+   */
+  private void checkConstraint(
+      SqlValidatorTable table,
+      SqlNode source,
+      RelDataType logicalTargetRowType) {
+    final ModifiableViewTable modifiableViewTable = table.unwrap(ModifiableViewTable.class);
+    if (modifiableViewTable != null && source instanceof SqlCall) {
+      final Map<Integer, RexNode> projectMap =
+          getConstraintForModifiableView(modifiableViewTable, logicalTargetRowType);
+      final List<SqlNode> values = ((SqlCall) source).getOperandList();
+      for (SqlNode row : values) {
+        for (Map.Entry<Integer, RexNode> entry : projectMap.entrySet()) {
+          Integer projectedOrdinal = null;
+          for (Integer ordinal : modifiableViewTable.getColumnMapping()) {
```

Comment:
If projectMap is a SortedMap, we could use a PriorityQueue here to improve the time complexity. wdyt?